src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,134 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import breeze.linalg.max
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+import scala.collection.mutable
+
+class Trie(normFunc: String => String = x => x) extends Serializable {
```

Comment:
I only just skimmed this but I don't see a reason this can't be immutable. `put` etc. can just return new instances

Reply:
The map is mutable to speed up the process of creation. Importing mutable is for mutable.Map. Otherwise, each put call would make a new copy of the internal map, and then runtime is more like n^2*(word length) than n*(word length)

Reply:
Performance would be a legit reason to make it mutable, however, did you benchmark it?
This http://docs.scala-lang.org/overviews/collections/performance-characteristics.html leads me to believe the O would be the same

Reply:
I made the change, scala seems to handle immutable trees really well.